### PR TITLE
bee: init at 0.1.0

### PR DESCRIPTION
**Pull request by paulperegud**
###### Motivation for this change
Bee is an official implementation of Ethereum's Swarm - distributed storage platform.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasj19** [comment]
Tested with `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 91825"` and works fine. You can check the third box in the description.

**Review by puzzlewolf — CHANGES_REQUESTED** [review]
Hi, thanks for your first PR :tada: 

Looks good, binary seems to work, just two small changes :)

Also, it's common to add a new maintainer in a separate commit. Please split your commit into two: `maintainers: add paulperegud` and `bee: init at 0.1.0`.

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -6052,6 +6052,12 @@
     githubId = 673857;
     name = "Pash Shocky";
   };
+  paulperegud = {
+    email = "paulperegud@gmail.com";
+    github = "paulperegud";
+    githubId = 247162;
+    name = "Pawel 'pepesza' Peregud";
+  };
```
Comment (kalbasit) [inline maintainers/maintainer-list.nix:6060]: This should be split into its own commit.

**Review by dasj19 — APPROVED** [review]

File: pkgs/applications/networking/bee/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  ];
+
+  vendorSha256 = "1a9fmf5c3asvylcb0ixlwb6jk2z5hzdaa6yffg1pjvq9rx92jbjy";
+
+  src = fetchFromGitHub {
+    owner = "ethersphere";
+    repo = "bee";
+    rev = "v${version}";
+    sha256 = "1iwwhs2zm8ml484sc0d7smmmp1x893zjafr1mnaswl8pxg0lnm2l";
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/bee/default.nix:18]:
```suggestion
  src = fetchFromGitHub {
    owner = "ethersphere";
    repo = "bee";
    rev = "v${version}";
    sha256 = "1iwwhs2zm8ml484sc0d7smmmp1x893zjafr1mnaswl8pxg0lnm2l";
  };

  vendorSha256 = "1a9fmf5c3asvylcb0ixlwb6jk2z5hzdaa6yffg1pjvq9rx92jbjy";

  subPackages = [ "cmd/bee" ];  
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 91825` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bee</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 91825` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bee</li>
  </ul>
</details>

**Comment by attila-lendvai** [comment]
bee is at version 0.4.1 at this point. updating to that would most probably fix this issue.

i'm new to nix, but i'm willing to try to update this PR and maybe even add a service to it. how shall i proceed?

shall i create a new PR and reference this one?

or can i get write access to this PR somehow and then i just update this one?

File: pkgs/applications/networking/bee/default.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, buildGoModule, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "bee";
+  version = "0.1.0";
```
Comment (attila-lendvai) [inline pkgs/applications/networking/bee/default.nix:5]:
```suggestion
  version = "0.4.1";
```

File: pkgs/applications/networking/bee/default.nix [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  src = fetchFromGitHub {
+    owner = "ethersphere";
+    repo = "bee";
+    rev = "v${version}";
+    sha256 = "1iwwhs2zm8ml484sc0d7smmmp1x893zjafr1mnaswl8pxg0lnm2l";
```
Comment (attila-lendvai) [inline pkgs/applications/networking/bee/default.nix:11]:
```suggestion
    sha256 = "1bmgbav52pcb5p7cgq9756512fzfqhjybyr0dv538plkqx47mpv7";
```

File: pkgs/applications/networking/bee/default.nix [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+  vendorSha256 = "1a9fmf5c3asvylcb0ixlwb6jk2z5hzdaa6yffg1pjvq9rx92jbjy";
```
Comment (attila-lendvai) [inline pkgs/applications/networking/bee/default.nix:14]:
```suggestion
  vendorSha256 = "0j393va4jrg9q3wlc9mgkbpgnn2w2s3k2hcn8phzj8d5fl4n4v2h";
```

**Review by attila-lendvai — CHANGES_REQUESTED** [review]
i'm very new to all this. i've changed the `~/.cache/nixpkgs-review/pr-91825/build.nix` file, but with my requested changes, it fails to build, and i don't know how to extract more information about the actual error:

```
$ nix build --no-link --keep-going --option build-use-sandbox relaxed -f /home/alendvai/.cache/nixpkgs-review/pr-91825/build.nix
builder for '/nix/store/dkxm5g3fb9pdfvmlcy1s22p8y8326q6a-bee-0.4.1.drv' failed with exit code 1; last 10 log lines:
  github.com/pelletier/go-toml
  github.com/spf13/afero/mem
  github.com/spf13/afero
  github.com/spf13/cast
  github.com/spf13/jwalterweatherman
  github.com/subosito/gotenv
  gopkg.in/ini.v1
  gopkg.in/yaml.v2
  github.com/spf13/viper
  golang.org/x/crypto/ssh/terminal
cannot build derivation '/nix/store/3s49s30yn4kgka6fishzhjb2s2lhls3b-env.drv': 1 dependencies couldn't be built
[0 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/3s49s30yn4kgka6fishzhjb2s2lhls3b-env.drv' failed
```

**Comment by paulperegud** [comment]
I'm also trying to create a 4.1 build basing on this PR. Debugging things that you are seeing. BTW - I'm also pretty new to nix :-)

**Comment by SuperSandro2000** [comment]
> bee is at version 0.4.1 at this point. updating to that would most probably fix this issue.
> 
> i'm new to nix, but i'm willing to try to update this PR and maybe even add a service to it. how shall i proceed?
> 
> shall i create a new PR and reference this one?
> 
> or can i get write access to this PR somehow and then i just update this one?

Yeah, create a new one and write in the description of it that it replaces this one.

**Comment by SuperSandro2000** [comment]
> i'm very new to all this. i've changed the `~/.cache/nixpkgs-review/pr-91825/build.nix` file, but with my requested changes, it fails to build, and i don't know how to extract more information about the actual error:
> 
> ```
> $ nix build --no-link --keep-going --option build-use-sandbox relaxed -f /home/alendvai/.cache/nixpkgs-review/pr-91825/build.nix
> builder for '/nix/store/dkxm5g3fb9pdfvmlcy1s22p8y8326q6a-bee-0.4.1.drv' failed with exit code 1; last 10 log lines:
>   github.com/pelletier/go-toml
>   github.com/spf13/afero/mem
>   github.com/spf13/afero
>   github.com/spf13/cast
>   github.com/spf13/jwalterweatherman
>   github.com/subosito/gotenv
>   gopkg.in/ini.v1
>   gopkg.in/yaml.v2
>   github.com/spf13/viper
>   golang.org/x/crypto/ssh/terminal
> cannot build derivation '/nix/store/3s49s30yn4kgka6fishzhjb2s2lhls3b-env.drv': 1 dependencies couldn't be built
> [0 built (1 failed), 0.0 MiB DL]
> error: build of '/nix/store/3s49s30yn4kgka6fishzhjb2s2lhls3b-env.drv' failed
> ```

take a look into the logs directory. There are the full logs.

**Comment by attila-lendvai** [comment]
so, the issue is that it builds fine, but fails with an assert at runtime:

```
$ ./results/bee/bin/bee version
panic: qtls.ConnectionState not compatible with tls.ConnectionState
```

the actual issue we are facing is discussed here: https://github.com/lucas-clemente/quic-go/issues/2614

according to this, it should work using `go_1_14`, but i couldn't find a way to specify the exact golang version to be used.

but on the other hand, if i check out the version from the bee repo, then i can build it using golang 1.15, which is strange to me.

```
$ go version
go version go1.15.5 linux/amd64

[nix-shell:~/workspace/ethereum/bee]$ git status
HEAD detached at v0.4.1

[nix-shell:~/workspace/ethereum/bee]$ ./dist/bee version
0.4.1-4b98b68
```

maybe it's because buildGoModule is not building with the exact same dependencies as a `make binary` in the checked out sources? it may be related to what's discussed here (long): https://github.com/NixOS/nixpkgs/issues/84826

**Comment by SuperSandro2000** [comment]
You can change the verison of go used by replacing buildGoModule wit heg buildGo114Module.

**Comment by attila-lendvai** [comment]
i have a reasonable default.nix for bee now that i think should work, but it doesn't compile, because of missing header files:

```
github.com/karalabe/usb
# github.com/karalabe/usb
vendor/github.com/karalabe/usb/hid_enabled.go:22:10: fatal error: ./hidapi/hidapi/hidapi.h: No such file or directory
   22 | #include "./hidapi/hidapi/hidapi.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

```
github.com/ethereum/go-ethereum/crypto/secp256k1
# github.com/ethereum/go-ethereum/crypto/secp256k1
vendor/github.com/ethereum/go-ethereum/crypto/secp256k1/curve.go:42:10: fatal error: libsecp256k1/include/secp256k1.h: No such file or directory
   42 | #include "libsecp256k1/include/secp256k1.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

my default.nix:

```
{ stdenv, buildGoModule, fetchFromGitHub, hidapi, secp256k1 }:

buildGoModule rec {
#buildGo114Module rec {
  pname = "bee";
  version = "0.4.1";

  #nativeBuildInputs = [ hidapi secp256k1 ];
  buildInputs = [ hidapi secp256k1 ];

  src = fetchFromGitHub {
    owner = "ethersphere";
    repo = "bee";
    rev = "v${version}";
    sha256 = "1bmgbav52pcb5p7cgq9756512fzfqhjybyr0dv538plkqx47mpv7";
  };

  vendorSha256 = "0j393va4jrg9q3wlc9mgkbpgnn2w2s3k2hcn8phzj8d5fl4n4v2h";

  subPackages = [ "cmd/bee" ];

  meta = with stdenv.lib; {
    homepage = "https://swarm-gateways.net/bzz:/docs.swarm.eth/";
    description = "Ethereum Swarm Bee";
    license = with licenses; [ bsd3 ];
    maintainers = with maintainers; [ paulperegud ];
  };
}
```

as you can see i have specified the dependencies with `buildInputs`, yet the .h files are not available. i've been searching and reading, but to no avail. AFAIU there are no separate `-dev` packages in the Nix universe.

this is also not a very useful error message for a newcomer like me:

```
$ nix-shell pkgs/applications/networking/bee/default.nix 
error: cannot auto-call a function that has an argument without a default value ('stdenv')
```

@SuperSandro2000 any hints on what to read/understand, or any examples to look at? i even searched github for `buildGoModule` and tried to find something similar. (sorry if i missed your answer on IRC)

**Comment by SuperSandro2000** [comment]
> AFAIU there are no separate -dev packages in the Nix universe.

The dev dependency is has .dev if there is one. buildInputs should be smart enough to choose that.

The hidapi header file is this ``/nix/store/hz6mghlv18dzfhn22dgll32jnzfly6sr-hidapi-0.10.1/include/hidapi/hidapi.h``

``./hidapi/hidapi/hidapi.h`` does not exist and may be a local header file. I think ``runVend = true;`` vendors those files.

Same with ``libsecp256k1/include/secp256k1.h``. The file in nix is ``/nix/store/yx1mgkm2x0ssfm20g10w4dd639gslksr-secp256k1-2020-08-16/include/secp256k1.h``.

You either need to change the includes or patch the includes.

> this is also not a very useful error message for a newcomer like me:

Nix 3.0 should improve this by a bit.

> nix-shell pkgs/applications/networking/bee/default.nix 

You can't do this. Go into the root of your nixpkgs clone and run ``nix-build -A bee``.

> nix-shell -E 'with import <nixpkgs> { }; callPackage ./pkgs/applications/networking/bee/default.nix { }'

That opens a nix-shell in that environment. You need to run some function like genericPhase or so. Don't remember the actual name right now.

**Comment by attila-lendvai** [comment]
@SuperSandro2000 thanks a whole lot!

i can now compile a bee and bee-unstable package. my branch has extensive changes on top of these two commits, so i'll open another PR that supersedes it once bee-clef and the bee service part is also packaged properly.

**Comment by attila-lendvai** [comment]
@paulperegud i'm making good progress with the packaging and created another (draft) PR: https://github.com/NixOS/nixpkgs/pull/109393

let me know if you wish to be included in the maintainers metadata. i've added you for now, but if you ask i can leave you out of this.

@SuperSandro2000 i'd appreciate any/all hints/reviews on the new PR.

**Comment by paulperegud** [comment]
@attila-lendvai Thank you! So I'm closing this one.
